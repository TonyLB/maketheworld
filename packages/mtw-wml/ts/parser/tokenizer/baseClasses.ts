import SourceStream from "./sourceStream";

export type TokenBase = {
    startIdx: number;
    endIdx: number;
}

export type TokenComment = {
    type: 'Comment'
} & TokenBase

export type TokenWhitespace = {
    type: 'Whitespace'
} & TokenBase

export type TokenExpressionValue = {
    type: 'ExpressionValue';
    value: string;
} & TokenBase

export type TokenKeyValue = {
    type: 'KeyValue';
    value: string;
} & TokenBase

export type TokenLiteralValue = {
    type: 'LiteralValue';
    value: string;
} & TokenBase

export type TokenBooleanProperty = {
    type: 'Property';
    isBoolean: true;
    key: string;
    value: boolean;
} & TokenBase

export type TokenEqualsProperty = {
    type: 'Property';
    isBoolean: false;
    key: string;
} & TokenBase

export type TokenProperty = TokenBooleanProperty | TokenEqualsProperty

export type TokenTagOpenBegin = {
    type: 'TagOpenBegin',
    tag: string;
} & TokenBase

export type TokenTagOpenEnd = {
    type: 'TagOpenEnd';
    selfClosing: boolean;
} & TokenBase

export type TokenTagClose = {
    type: 'TagClose';
} & TokenBase

export type TokenDescription = {
    type: 'Description';
    value: string;
} & TokenBase

export type TokenError = {
    type: 'Error';
    message: string;
} & TokenBase

export class TokenizeException extends Error {
    startIdx: number
    endIdx: number
    constructor(message: string, startIdx: number, endIdx: number) {
        super(message)
        this.name = 'TokenizeException'
        this.startIdx = startIdx
        this.endIdx = endIdx
    }
}

export type Token = TokenComment |
    TokenWhitespace |
    TokenExpressionValue |
    TokenKeyValue |
    TokenLiteralValue |
    TokenProperty |
    TokenTagOpenBegin |
    TokenTagOpenEnd |
    TokenTagClose |
    TokenDescription |
    TokenError

export type Tokenizer<T extends TokenBase> = {
    (sourceStream: SourceStream): T | undefined
}

export const isTokenError = (item: TokenBase & { type: string }): item is TokenError => (item.type === 'Error')
export const isTokenProperty = (item: Token): item is TokenProperty => (item.type === 'Property')
export const isTokenKeyValue = (item: Token | boolean): item is TokenKeyValue => (typeof item === 'object' && item.type === 'KeyValue')
export const isTokenValue = (item: Token): item is (TokenKeyValue | TokenLiteralValue | TokenExpressionValue) => (['KeyValue', 'LiteralValue', 'ExpressionValue'].includes(item.type))
export const isTokenPropertyOrValue = (item: Token): item is (TokenProperty | TokenKeyValue | TokenLiteralValue | TokenExpressionValue) => (isTokenProperty(item) || isTokenValue(item))

export const isTokenWhitespace = (item: Token): item is TokenWhitespace => (item.type === 'Whitespace')
export const isTokenComment = (item: Token): item is TokenComment => (item.type === 'Comment')
export const isTokenDescription = (item: Token): item is TokenDescription => (item.type === 'Description')
export const isLegalBareToken = (item: Token): item is (TokenWhitespace | TokenComment | TokenDescription) => (['Whitespace', 'Comment', 'Description'].includes(item.type))
