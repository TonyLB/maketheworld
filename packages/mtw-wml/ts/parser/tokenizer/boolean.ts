import {
    Tokenizer,
    TokenBooleanProperty
} from "./baseClasses"

export const booleanPropertyTokenizer: Tokenizer<TokenBooleanProperty> = (sourceStream) => {
    const firstChar = sourceStream.lookAhead(1)
    if (firstChar.match(/[A-Za-z\!]/)) {
        const startIdx = sourceStream.position
        let returnValue = firstChar === '!' ? sourceStream.consume(1) : ''
        let key = ''
        while(sourceStream.lookAhead(1).match(/[A-Za-z]/)) {
            key = key + sourceStream.consume(1)
        }
        returnValue = returnValue + key
        return {
            type: 'BooleanValue',
            source: returnValue,
            value: firstChar !== '!',
            key,
            startIdx,
            endIdx: sourceStream.position - 1
        }
    }
    else {
        return undefined
    }
}

export default booleanPropertyTokenizer
