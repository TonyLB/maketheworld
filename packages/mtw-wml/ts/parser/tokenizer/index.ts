import {
    Token,
    TokenDescription,
    TokenTagClose,
    TokenTagOpenBegin,
    TokenTagOpenEnd,
    TokenWhitespace,
    TokenProperty,
    TokenLiteralValue,
    TokenKeyValue,
    TokenExpressionValue,
    TokenComment,
    TokenizeException
} from "./baseClasses"
import expressionValueTokenizer from "./expression"
import keyValueTokenizer from "./key"
import literalValueTokenizer from "./literal"
import SourceStream from "./sourceStream"
import tagCloseTokenizer from "./tagClose"
import descriptionTokenizer from "./description"

import { tagOpenBeginTokenizer, tagOpenEndTokenizer } from "./tagOpen"
import tagPropertyTokenizer from "./tagProperty"
import whiteSpaceTokenizer from "./whiteSpace"
import { checkSubTokenizers } from "./utils"
import commentTokenizer from "./comments"

enum TokenizerMode {
    Contents,
    Tag
}

export const tokenizer = (sourceStream: SourceStream): Token[] => {
    let mode: TokenizerMode = TokenizerMode.Contents
    let returnValue: Token[] = []

    let currentDescription: TokenDescription | undefined
    let currentWhitespace: TokenWhitespace | undefined
    while(!sourceStream.isEndOfSource) {
        if (mode === TokenizerMode.Contents) {
            //
            // TODO: Parse and buffer Descriptions, merging together whitespace as you go,
            // and checking for \-escaped characters, instead of regarding other data as an error.
            //
            const checkSubTokens = checkSubTokenizers<TokenComment | TokenWhitespace | TokenTagClose | TokenTagOpenBegin | TokenDescription>({
                sourceStream,
                subTokenizers: [commentTokenizer, whiteSpaceTokenizer, tagCloseTokenizer, tagOpenBeginTokenizer, descriptionTokenizer],
                callback: (token) => {
                    if (token.type === 'TagClose' || token.type === 'TagOpenBegin' || token.type === 'Comment') {
                        if (currentDescription) {
                            returnValue.push(currentDescription)
                            currentDescription = undefined
                        }
                        if (currentWhitespace) {
                            returnValue.push(currentWhitespace)
                            currentWhitespace = undefined
                        }
                        returnValue.push(token)
                        if (token.type === 'TagOpenBegin') {
                            mode = TokenizerMode.Tag
                        }
                    }
                    if (token.type === 'Whitespace') {
                        if (currentDescription) {
                            currentDescription = {
                                type: 'Description',
                                startIdx: currentDescription.startIdx,
                                endIdx: token.endIdx,
                                value: sourceStream.source.slice(currentDescription.startIdx, token.endIdx)
                            }
                        }
                        else {
                            currentWhitespace = token
                        }
                    }
                    if (token.type === 'Description') {
                        if (currentDescription) {
                            currentDescription = {
                                type: 'Description',
                                startIdx: currentDescription.startIdx,
                                endIdx: token.endIdx,
                                value: sourceStream.source.slice(currentDescription.startIdx, token.endIdx)
                            }
                        }
                        else if (currentWhitespace) {
                            currentDescription = {
                                type: 'Description',
                                startIdx: currentWhitespace.startIdx,
                                endIdx: token.endIdx,
                                value: sourceStream.source.slice(currentWhitespace.startIdx, token.endIdx)
                            }
                            currentWhitespace = undefined
                        }
                        else {
                            currentDescription = token
                        }
                    }
                }
            })
            if (checkSubTokens) {
                continue
            }
            return [{
                type: 'Error',
                startIdx: sourceStream.position,
                endIdx: sourceStream.position,
                message: 'Unexpected token'
            }]
        }
        else {
            const checkSubTokens = checkSubTokenizers<TokenTagOpenEnd | TokenComment | TokenWhitespace | TokenProperty>({
                sourceStream,
                subTokenizers: [tagOpenEndTokenizer, commentTokenizer, whiteSpaceTokenizer, tagPropertyTokenizer],
                callback: (token) => {
                    returnValue.push(token)
                    if (token.type === 'TagOpenEnd') {
                        mode = TokenizerMode.Contents
                    }
                    if (token.type === 'Property' && !token.isBoolean) {
                        //
                        // Consume any whitespace first...
                        //
                        const checkWhitespace = whiteSpaceTokenizer(sourceStream)
                        if (checkWhitespace) {
                            returnValue.push(checkWhitespace)
                        }
                        //
                        // ... and then confirm that a value follows the key
                        //
                        const checkValue = checkSubTokenizers<TokenLiteralValue | TokenKeyValue | TokenExpressionValue>({
                            sourceStream,
                            subTokenizers: [literalValueTokenizer, keyValueTokenizer, expressionValueTokenizer],
                            callback: (token) => {
                                returnValue.push(token)
                            }
                        })
                        if (!checkValue) {
                            throw new TokenizeException('Unexpected token', sourceStream.position, sourceStream.position)
                        }
                    }
                }
            })
            if (!checkSubTokens) {
                throw new TokenizeException('Unexpected token', sourceStream.position, sourceStream.position)
            }
        }
    }
    return returnValue
}

export default tokenizer
