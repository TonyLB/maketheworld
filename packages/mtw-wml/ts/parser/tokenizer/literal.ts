import {
    Tokenizer,
    TokenLiteralValue,
    TokenizeException
} from "./baseClasses"

export const expressionStringLiteralSubTokenizer: Tokenizer<TokenLiteralValue> = (sourceStream) => {
    const firstChar = sourceStream.lookAhead(1)
    if ([`'`, `"`].includes(firstChar)) {
        const startIdx = sourceStream.position
        sourceStream.consume(1)
        const valueStartIdx = sourceStream.position
        while(!sourceStream.lookAhead(firstChar)) {
            const nextChar = sourceStream.lookAhead(1)
            if (nextChar === '\\') {
                sourceStream.consume(2)
            }
            else {
                sourceStream.consume(1)
            }
            if (sourceStream.isEndOfSource) {
                throw new TokenizeException('Unbounded string literal', startIdx, sourceStream.position - 1)
            }
        }
        const value = sourceStream.source.slice(valueStartIdx, sourceStream.position)
        sourceStream.consume(1)
        return {
            type: 'LiteralValue',
            value,
            startIdx,
            endIdx: sourceStream.position - 1
        }
    }
    else {
        return undefined
    }
}

export const literalValueTokenizer: Tokenizer<TokenLiteralValue> = (sourceStream) => {
    if (sourceStream.lookAhead('"')) {
        return expressionStringLiteralSubTokenizer(sourceStream)
    }
    return undefined
}

export default literalValueTokenizer
