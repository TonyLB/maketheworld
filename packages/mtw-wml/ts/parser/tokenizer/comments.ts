import { Tokenizer, TokenComment } from "./baseClasses"

export const commentTokenizer: Tokenizer<TokenComment> = (sourceStream) => {
    if (sourceStream.lookAhead('//')) {
        const nextInstance = sourceStream.nextInstance(['\n'])
        const startIdx = sourceStream.position
        if (nextInstance > 1) {
            sourceStream.consume(nextInstance)
        }
        else {
            sourceStream.consume(sourceStream.source.length - sourceStream.position)
        }
        const endIdx = sourceStream.position - 1
        return {
            type: 'Comment',
            startIdx,
            endIdx
        }
    }
    if (sourceStream.lookAhead('/*')) {
        const nextInstance = sourceStream.nextInstance(['*/'])
        const startIdx = sourceStream.position
        if (nextInstance > 1) {
            sourceStream.consume(nextInstance + 2)
        }
        else {
            sourceStream.consume(sourceStream.source.length - sourceStream.position)
        }
        const endIdx = sourceStream.position - 1
        if (nextInstance === -1) {
            return {
                type: 'Error',
                message: 'Unbounded comment',
                startIdx,
                endIdx
            }
        }
        else {
            return {
                type: 'Comment',
                startIdx,
                endIdx
            }
        }
    }
    return undefined
}

export default commentTokenizer
