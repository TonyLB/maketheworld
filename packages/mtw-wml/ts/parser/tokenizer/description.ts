import { Tokenizer, TokenDescription } from "./baseClasses"

export const descriptionTokenizer: Tokenizer<TokenDescription> = (sourceStream) => {
    const startIdx = sourceStream.position
    let value = ''
    while (![' ', '\n', '\t', '\r', '<'].includes(sourceStream.lookAhead(1))) {
        if (['/*', '//'].includes(sourceStream.lookAhead(2))) {
            break
        }
        if (sourceStream.isEndOfSource) {
            break
        }
        if (sourceStream.lookAhead(1) === '\\') {
            sourceStream.consume(1)
        }
        value = `${value}${sourceStream.consume(1)}`
    }
    const endIdx = sourceStream.position
    if (startIdx === endIdx) {
        return undefined
    }
    else {
        return {
            type: 'Description',
            startIdx,
            endIdx,
            value
        }
    }
}

export default descriptionTokenizer
