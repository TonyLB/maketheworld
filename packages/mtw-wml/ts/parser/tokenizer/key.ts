import { Tokenizer, TokenKeyValue, TokenizeException } from "./baseClasses"
import whiteSpaceTokenizer from "./whiteSpace"
import { checkSubTokenizers } from "./utils"

export const keyValueTokenizer: Tokenizer<TokenKeyValue> = (sourceStream) => {
    const firstChar = sourceStream.lookAhead(1)
    if (firstChar === '(') {
        const startIdx = sourceStream.position
        sourceStream.consume(1)
        checkSubTokenizers({
            sourceStream,
            subTokenizers: [whiteSpaceTokenizer],
            callback: (token) => {}
        })
        const valueStartIdx = sourceStream.position
        if (sourceStream.lookAhead(1).match(/[A-Za-z\_]/)) {
            sourceStream.consume(1)
            while(sourceStream.lookAhead(1).match(/[A-Za-z0-9\_]/)) {
                sourceStream.consume(1)
            }    
        }
        const value = sourceStream.source.slice(valueStartIdx, sourceStream.position)
        checkSubTokenizers({
            sourceStream,
            subTokenizers: [whiteSpaceTokenizer],
            callback: (token) => {}
        })
        if (!sourceStream.lookAhead(')')) {
            throw new TokenizeException('Unexpected token', sourceStream.position, sourceStream.position)
        }
        sourceStream.consume(1)
        if (sourceStream.position !== startIdx) {
            return {
                type: 'KeyValue',
                value,
                startIdx,
                endIdx: sourceStream.position - 1
            }
        }
    }
    return undefined
}

export default keyValueTokenizer
