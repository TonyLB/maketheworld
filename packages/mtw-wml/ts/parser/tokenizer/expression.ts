import {
    TokenBase,
    Tokenizer,
    TokenExpressionValue
} from "./baseClasses"
import { expressionStringLiteralSubTokenizer } from "./literal"
import { checkSubTokenizer } from "./utils"

type TokenTemplateString = {
    type: 'TemplateString'
} & TokenBase

export const expressionTemplateStringSubTokenizer: Tokenizer<TokenTemplateString> = (sourceStream) => {
    const firstChar = sourceStream.lookAhead(1)
    if (firstChar === "`") {
        const startIdx = sourceStream.position
        let returnValue = sourceStream.consume(1)
        while(!sourceStream.lookAhead(firstChar)) {
            const nextChar = sourceStream.lookAhead(1)
            if (nextChar === '\\') {
                returnValue = returnValue + sourceStream.consume(2)
            }
            else {
                returnValue = returnValue + sourceStream.consume(1)
                if (nextChar === '$') {
                    const checkExpressionToken = checkSubTokenizer({
                        currentBuffer: returnValue,
                        startIdx,
                        subTokenizer: expressionValueTokenizer,
                        sourceStream
                    })
                    if (checkExpressionToken.error) {
                        return checkExpressionToken.error
                    }
                    else if (checkExpressionToken.returnBuffer) {
                        returnValue = checkExpressionToken.returnBuffer
                    }
                    else {
                        returnValue = returnValue + sourceStream.consume(1)
                    }
                }
            }
            if (sourceStream.isEndOfSource) {
                return {
                    type: 'Error',
                    message: 'Unbounded template string',
                    source: returnValue,
                    startIdx,
                    endIdx: sourceStream.position - 1
                }
            }
        }
        returnValue = returnValue + sourceStream.consume(1)
        return {
            type: 'TemplateString',
            source: returnValue,
            startIdx,
            endIdx: sourceStream.position - 1
        }
    }
    else {
        return undefined
    }
}

export const expressionValueTokenizer: Tokenizer<TokenExpressionValue> = (sourceStream) => {
    if (!sourceStream.lookAhead('{')) {
        return undefined
    }
    let returnValue = ''
    const startIdx = sourceStream.position
    returnValue = returnValue + sourceStream.consume(1)
    while(!sourceStream.lookAhead('}')) {
        const checkLiteral = checkSubTokenizer({
            currentBuffer: returnValue,
            startIdx,
            subTokenizer: expressionStringLiteralSubTokenizer,
            sourceStream
        })
        if (checkLiteral.error) {
            return checkLiteral.error
        }
        if (checkLiteral.returnBuffer) {
            returnValue = checkLiteral.returnBuffer
            continue
        }
        const checkTemplate = checkSubTokenizer({
            currentBuffer: returnValue,
            startIdx,
            subTokenizer: expressionTemplateStringSubTokenizer,
            sourceStream
        })
        if (checkTemplate.error) {
            return checkTemplate.error
        }
        if (checkTemplate.returnBuffer) {
            returnValue = checkTemplate.returnBuffer
            continue
        }
        const checkExpression = checkSubTokenizer({
            currentBuffer: returnValue,
            startIdx,
            subTokenizer: expressionValueTokenizer,
            sourceStream
        })
        if (checkExpression.error) {
            return checkExpression.error
        }
        if (checkExpression.returnBuffer) {
            returnValue = checkExpression.returnBuffer
            continue
        }
        returnValue = returnValue + sourceStream.consume(1)
    }
    returnValue = returnValue + sourceStream.consume(1)
    return {
        type: 'ExpressionValue',
        source: returnValue,
        startIdx,
        endIdx: sourceStream.position - 1
    }
}

export default expressionValueTokenizer
