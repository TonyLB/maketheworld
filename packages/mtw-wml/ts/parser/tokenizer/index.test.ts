import SourceStream from './sourceStream'
import tokenizer from '.'
import { TokenizeException } from './baseClasses'

describe('tokenizer', () => {
    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
    })

    it('should return empty list on empty source', () => {
        const testStream = new SourceStream('')
        expect(tokenizer(testStream)).toEqual([])
    })
    it('should tokenize a single self-closing tag', () => {
        const testStream = new SourceStream('<Room key=(ABC) />')
        expect(tokenizer(testStream)).toMatchSnapshot()
    })
    it('should tokenize a single non-self-closing tag', () => {
        const testStream = new SourceStream('<Room key=(ABC)></Room>')
        expect(tokenizer(testStream)).toMatchSnapshot()
    })
    it('should tokenize nested tags', () => {
        const testSource = `<Asset key=(Test)>
            <Room global key=(ABC)>
                <Feature key=(DEF) />
            </Room>
        </Asset>`
        const testStream = new SourceStream(testSource)
        expect(tokenizer(testStream)).toMatchSnapshot()
    })
    it('should reject tag property with no value specified', () => {
        const testStream = new SourceStream('<Room key= >')
        expect(() => {
            tokenizer(testStream)
        }).toThrow(TokenizeException)
    })
    it('should tokenize descriptions', () => {
        const testSource = `<Asset key=(Test)>
            <Room key=(ABC)>
                <Description>
                    Test description
                    <Link key=(DEF)>link text</Link>
                    End of test
                </Description>
            </Room>
        </Asset>`
        const testStream = new SourceStream(testSource)
        expect(tokenizer(testStream)).toMatchSnapshot()
    })
    it('should tokenize comments', () => {
        const testSource = `<Asset key=(Test)>
            <Room /* Comment */ key=(ABC)>
                <Description>
                    Test description//Comment
                    /* Comment */
                    End of test
                </Description>
            </Room>
        </Asset>`
        const testStream = new SourceStream(testSource)
        expect(tokenizer(testStream)).toMatchSnapshot()
    })
    xit('should perform quickly (only activate when performance-tuning)', () => {
        const testSource = `<Asset key=(Test)>
            <Room key=(ABC)>
                <Feature key=(DEF) />
            </Room>
        </Asset>`
        const testStream = new SourceStream(testSource)
        const start = Date.now()
        const returnValue = tokenizer(testStream)
        const duration = Date.now() - start
        expect(duration < 10).toBe(true)
    })

})
