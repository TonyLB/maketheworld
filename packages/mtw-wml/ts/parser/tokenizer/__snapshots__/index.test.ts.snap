// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenizer should tokenize a single non-self-closing tag 1`] = `
Array [
  Object {
    "endIdx": 4,
    "startIdx": 0,
    "tag": "Room",
    "type": "TagOpenBegin",
  },
  Object {
    "endIdx": 5,
    "startIdx": 5,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 9,
    "isBoolean": false,
    "key": "key",
    "startIdx": 6,
    "type": "Property",
  },
  Object {
    "endIdx": 14,
    "startIdx": 10,
    "type": "KeyValue",
    "value": "ABC",
  },
  Object {
    "endIdx": 15,
    "selfClosing": false,
    "startIdx": 15,
    "type": "TagOpenEnd",
  },
  Object {
    "endIdx": 22,
    "startIdx": 16,
    "tag": "Room",
    "type": "TagClose",
  },
]
`;

exports[`tokenizer should tokenize a single self-closing tag 1`] = `
Array [
  Object {
    "endIdx": 4,
    "startIdx": 0,
    "tag": "Room",
    "type": "TagOpenBegin",
  },
  Object {
    "endIdx": 5,
    "startIdx": 5,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 9,
    "isBoolean": false,
    "key": "key",
    "startIdx": 6,
    "type": "Property",
  },
  Object {
    "endIdx": 14,
    "startIdx": 10,
    "type": "KeyValue",
    "value": "ABC",
  },
  Object {
    "endIdx": 15,
    "startIdx": 15,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 17,
    "selfClosing": true,
    "startIdx": 16,
    "type": "TagOpenEnd",
  },
]
`;

exports[`tokenizer should tokenize comments 1`] = `
Array [
  Object {
    "endIdx": 5,
    "startIdx": 0,
    "tag": "Asset",
    "type": "TagOpenBegin",
  },
  Object {
    "endIdx": 6,
    "startIdx": 6,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 10,
    "isBoolean": false,
    "key": "key",
    "startIdx": 7,
    "type": "Property",
  },
  Object {
    "endIdx": 16,
    "startIdx": 11,
    "type": "KeyValue",
    "value": "Test",
  },
  Object {
    "endIdx": 17,
    "selfClosing": false,
    "startIdx": 17,
    "type": "TagOpenEnd",
  },
  Object {
    "endIdx": 30,
    "startIdx": 18,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 35,
    "startIdx": 31,
    "tag": "Room",
    "type": "TagOpenBegin",
  },
  Object {
    "endIdx": 36,
    "startIdx": 36,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 49,
    "startIdx": 37,
    "type": "Comment",
  },
  Object {
    "endIdx": 50,
    "startIdx": 50,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 54,
    "isBoolean": false,
    "key": "key",
    "startIdx": 51,
    "type": "Property",
  },
  Object {
    "endIdx": 59,
    "startIdx": 55,
    "type": "KeyValue",
    "value": "ABC",
  },
  Object {
    "endIdx": 60,
    "selfClosing": false,
    "startIdx": 60,
    "type": "TagOpenEnd",
  },
  Object {
    "endIdx": 77,
    "startIdx": 61,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 89,
    "startIdx": 78,
    "tag": "Description",
    "type": "TagOpenBegin",
  },
  Object {
    "endIdx": 90,
    "selfClosing": false,
    "startIdx": 90,
    "type": "TagOpenEnd",
  },
  Object {
    "endIdx": 128,
    "startIdx": 91,
    "type": "Description",
    "value": "
                    Test description",
  },
  Object {
    "endIdx": 136,
    "startIdx": 128,
    "type": "Comment",
  },
  Object {
    "endIdx": 157,
    "startIdx": 137,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 170,
    "startIdx": 158,
    "type": "Comment",
  },
  Object {
    "endIdx": 219,
    "startIdx": 171,
    "type": "Description",
    "value": "
                    End of test
               ",
  },
  Object {
    "endIdx": 233,
    "startIdx": 220,
    "tag": "Description",
    "type": "TagClose",
  },
  Object {
    "endIdx": 246,
    "startIdx": 234,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 253,
    "startIdx": 247,
    "tag": "Room",
    "type": "TagClose",
  },
  Object {
    "endIdx": 262,
    "startIdx": 254,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 270,
    "startIdx": 263,
    "tag": "Asset",
    "type": "TagClose",
  },
]
`;

exports[`tokenizer should tokenize descriptions 1`] = `
Array [
  Object {
    "endIdx": 5,
    "startIdx": 0,
    "tag": "Asset",
    "type": "TagOpenBegin",
  },
  Object {
    "endIdx": 6,
    "startIdx": 6,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 10,
    "isBoolean": false,
    "key": "key",
    "startIdx": 7,
    "type": "Property",
  },
  Object {
    "endIdx": 16,
    "startIdx": 11,
    "type": "KeyValue",
    "value": "Test",
  },
  Object {
    "endIdx": 17,
    "selfClosing": false,
    "startIdx": 17,
    "type": "TagOpenEnd",
  },
  Object {
    "endIdx": 30,
    "startIdx": 18,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 35,
    "startIdx": 31,
    "tag": "Room",
    "type": "TagOpenBegin",
  },
  Object {
    "endIdx": 36,
    "startIdx": 36,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 40,
    "isBoolean": false,
    "key": "key",
    "startIdx": 37,
    "type": "Property",
  },
  Object {
    "endIdx": 45,
    "startIdx": 41,
    "type": "KeyValue",
    "value": "ABC",
  },
  Object {
    "endIdx": 46,
    "selfClosing": false,
    "startIdx": 46,
    "type": "TagOpenEnd",
  },
  Object {
    "endIdx": 63,
    "startIdx": 47,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 75,
    "startIdx": 64,
    "tag": "Description",
    "type": "TagOpenBegin",
  },
  Object {
    "endIdx": 76,
    "selfClosing": false,
    "startIdx": 76,
    "type": "TagOpenEnd",
  },
  Object {
    "endIdx": 134,
    "startIdx": 77,
    "type": "Description",
    "value": "
                    Test description
                   ",
  },
  Object {
    "endIdx": 139,
    "startIdx": 135,
    "tag": "Link",
    "type": "TagOpenBegin",
  },
  Object {
    "endIdx": 140,
    "startIdx": 140,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 144,
    "isBoolean": false,
    "key": "key",
    "startIdx": 141,
    "type": "Property",
  },
  Object {
    "endIdx": 149,
    "startIdx": 145,
    "type": "KeyValue",
    "value": "DEF",
  },
  Object {
    "endIdx": 150,
    "selfClosing": false,
    "startIdx": 150,
    "type": "TagOpenEnd",
  },
  Object {
    "endIdx": 160,
    "startIdx": 151,
    "type": "Description",
    "value": "link text",
  },
  Object {
    "endIdx": 166,
    "startIdx": 160,
    "tag": "Link",
    "type": "TagClose",
  },
  Object {
    "endIdx": 215,
    "startIdx": 167,
    "type": "Description",
    "value": "
                    End of test
               ",
  },
  Object {
    "endIdx": 229,
    "startIdx": 216,
    "tag": "Description",
    "type": "TagClose",
  },
  Object {
    "endIdx": 242,
    "startIdx": 230,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 249,
    "startIdx": 243,
    "tag": "Room",
    "type": "TagClose",
  },
  Object {
    "endIdx": 258,
    "startIdx": 250,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 266,
    "startIdx": 259,
    "tag": "Asset",
    "type": "TagClose",
  },
]
`;

exports[`tokenizer should tokenize nested tags 1`] = `
Array [
  Object {
    "endIdx": 5,
    "startIdx": 0,
    "tag": "Asset",
    "type": "TagOpenBegin",
  },
  Object {
    "endIdx": 6,
    "startIdx": 6,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 10,
    "isBoolean": false,
    "key": "key",
    "startIdx": 7,
    "type": "Property",
  },
  Object {
    "endIdx": 16,
    "startIdx": 11,
    "type": "KeyValue",
    "value": "Test",
  },
  Object {
    "endIdx": 17,
    "selfClosing": false,
    "startIdx": 17,
    "type": "TagOpenEnd",
  },
  Object {
    "endIdx": 30,
    "startIdx": 18,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 35,
    "startIdx": 31,
    "tag": "Room",
    "type": "TagOpenBegin",
  },
  Object {
    "endIdx": 36,
    "startIdx": 36,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 43,
    "isBoolean": true,
    "key": "global",
    "startIdx": 37,
    "type": "Property",
    "value": true,
  },
  Object {
    "endIdx": 47,
    "isBoolean": false,
    "key": "key",
    "startIdx": 44,
    "type": "Property",
  },
  Object {
    "endIdx": 52,
    "startIdx": 48,
    "type": "KeyValue",
    "value": "ABC",
  },
  Object {
    "endIdx": 53,
    "selfClosing": false,
    "startIdx": 53,
    "type": "TagOpenEnd",
  },
  Object {
    "endIdx": 70,
    "startIdx": 54,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 78,
    "startIdx": 71,
    "tag": "Feature",
    "type": "TagOpenBegin",
  },
  Object {
    "endIdx": 79,
    "startIdx": 79,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 83,
    "isBoolean": false,
    "key": "key",
    "startIdx": 80,
    "type": "Property",
  },
  Object {
    "endIdx": 88,
    "startIdx": 84,
    "type": "KeyValue",
    "value": "DEF",
  },
  Object {
    "endIdx": 89,
    "startIdx": 89,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 91,
    "selfClosing": true,
    "startIdx": 90,
    "type": "TagOpenEnd",
  },
  Object {
    "endIdx": 104,
    "startIdx": 92,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 111,
    "startIdx": 105,
    "tag": "Room",
    "type": "TagClose",
  },
  Object {
    "endIdx": 120,
    "startIdx": 112,
    "type": "Whitespace",
  },
  Object {
    "endIdx": 128,
    "startIdx": 121,
    "tag": "Asset",
    "type": "TagClose",
  },
]
`;
